barter/examples/engine_with_custom_market_data.rs

```diff
@@ -0,0 +1,259 @@
+use barter::{
+    engine::{
+        audit::Audit,
+        clock::{EngineClock, LiveClock},
+        command::Command,
+        run,
+        state::{
+            instrument::{filter::InstrumentFilter, market_data::MarketDataState},
+            trading::TradingState,
+            EngineState,
+        },
+        Engine, Processor,
+    },
+    execution::builder::ExecutionBuilder,
+    logging::init_logging,
+    risk::{DefaultRiskManager, DefaultRiskManagerState},
+    statistic::time::Daily,
+    strategy::{DefaultStrategy, DefaultStrategyState},
+    EngineEvent,
+};
+use barter_data::{
+    books::OrderBook,
+    event::{DataKind, MarketEvent},
+    streams::{
+        builder::dynamic::indexed::init_indexed_multi_exchange_market_stream,
+        reconnect::stream::ReconnectingStream,
+    },
+    subscription::SubKind,
+};
+use barter_execution::{balance::Balance, client::mock::MockExecutionConfig};
+use barter_instrument::{
+    exchange::ExchangeId,
+    index::IndexedInstruments,
+    instrument::{
+        kind::InstrumentKind,
+        spec::{
+            InstrumentSpec, InstrumentSpecNotional, InstrumentSpecPrice, InstrumentSpecQuantity,
+            OrderQuantityUnits,
+        },
+        Instrument,
+    },
+    Underlying,
+};
+use barter_integration::channel::{mpsc_unbounded, ChannelTxDroppable, Tx};
+use fnv::FnvHashMap;
+use futures::StreamExt;
+use rust_decimal::Decimal;
+use rust_decimal_macros::dec;
+use tracing::debug;
+
+const EXCHANGE: ExchangeId = ExchangeId::BinanceSpot;
+const RISK_FREE_RETURN: Decimal = dec!(0.05);
+const MOCK_EXCHANGE_ROUND_TRIP_LATENCY_MS: u64 = 100;
+const MOCK_EXCHANGE_FEES_PERCENT: Decimal = dec!(0.05);
+const STARTING_BALANCE_USDT: Balance = Balance {
+    total: dec!(10_000.0),
+    free: dec!(10_000.0),
+};
+const STARTING_BALANCE_BTC: Balance = Balance {
+    total: dec!(0.1),
+    free: dec!(0.1),
+};
+const STARTING_BALANCE_ETH: Balance = Balance {
+    total: dec!(1.0),
+    free: dec!(1.0),
+};
+const STARTING_BALANCE_SOL: Balance = Balance {
+    total: dec!(10.0),
+    free: dec!(10.0),
+};
+
+/// Market data state that holds a full orderbook. This state can later be used
+/// in a strategy to make decisions on.
+#[derive(Debug, Default, Clone)]
+struct MarketData {
+    pub book: OrderBook,
```

Comment:
Although this is a valid approach I think it's better to have the OrderBook managed on a seperate thread since it could be resource intensive. 

We don't want to block the main engine thread with a bunch of order book management tasks. 

thoughts?

Reply:
ie/     pub book: Arc<RwLock<OrderBook>>,